File: topos/cli/commands/refactor.py
"""``topos refactor`` — unified refactoring-guidance suite.

Three subcommands sharing one output shape (ranked hotspot table + suggested
action), each surfacing a different structural-analysis engine from the
Methods Upgrade milestone:

- ``cycles``: cycle-basis extraction on the CFG (issue #83).
- ``dependencies``: balanced Forman curvature on the MDG (issue #84).
- ``process``: directed Forman-Ricci curvature on GitNexus process
  graphs (issue #86).

All three are advisory only — none of this affects ``topos evaluate``.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING

import click

if TYPE_CHECKING:
    from topos.mcp.schemas import RefactorHotspot


@click.group()
def refactor() -> None:
    """Actionable refactor suggestions: cycles, dependencies, process choke points."""


def _print_hotspots(
    title: str, hotspots: list[RefactorHotspot], output_json: bool
) -> None:
    if output_json:
        click.echo(json.dumps([h.model_dump() for h in hotspots], indent=2))
        return

    click.echo(title)
    if not hotspots:
        click.echo("  (none found)")
        return
    for h in hotspots:
        location = h.filepath
        if h.line_start is not None:
            location += f":{h.line_start}"
            if h.line_end is not None and h.line_end != h.line_start:
                location += f"-{h.line_end}"
        click.echo(f"  [{h.kind}] {h.label}  ({location})  score={h.score:.3f}")
        click.echo(f"    -> {h.suggestion}")


@refactor.command("cycles")
@click.argument("path", type=click.Path(exists=True, dir_okay=False))
@click.option(
    "--json", "output_json_flag", is_flag=True, help="Output results as JSON."
)
@click.option(
    "--max-cycles",
    default=5,
    show_default=True,
    type=int,
    help="Maximum cycle hotspots to display.",
)
def cycles_cmd(path: str, output_json_flag: bool, max_cycles: int) -> None:
    """List CFG cycle generators mapped to source line ranges (issue #83)."""
    from topos.core.morphism import ProgramMorphism
    from topos.functors.probes.cfg.homology import calculate_cycle_basis
    from topos.mcp.schemas import RefactorHotspot

    morphism = ProgramMorphism.from_file(path)
    cfg = morphism.build_cfg()
    if cfg is None:
        click.echo(f"Error: could not build a control-flow graph for {path}", err=True)
        raise SystemExit(1)

    result = calculate_cycle_basis(cfg)

    def _span(cycle) -> int:
        if cycle.start_line is None or cycle.end_line is None:
            return 0
        return cycle.end_line - cycle.start_line

    ranked = sorted(result.cycles, key=_span, reverse=True)[:max_cycles]
    hotspots = [
        RefactorHotspot(
            kind="cycle",
            label=f"cycle over blocks {cycle.block_ids}",
            filepath=path,
            line_start=cycle.start_line,
            line_end=cycle.end_line,
            score=float(_span(cycle)),
            suggestion=(
                "Extract this loop/branch body into its own function to "
                "isolate the cycle and shrink cyclomatic complexity."
            ),
        )
        for cycle in ranked
    ]
    title = f"Cycle hotspots (betti_1={result.betti_1}):"
    _print_hotspots(title, hotspots, output_json_flag)


@refactor.command("dependencies")
@click.argument("path", type=click.Path(exists=True, dir_okay=False))
@click.option(
    "--gitnexus-dir",
    type=click.Path(exists=True, file_okay=False),
    default=None,
    help="Path to a .gitnexus/ directory (auto-detected from cwd when omitted).",
)
@click.option(
    "--json", "output_json_flag", is_flag=True, help="Output results as JSON."
)
@click.option(
    "--max-targets",
    default=5,
    show_default=True,
    type=int,
    help="Maximum dependency hotspots to display.",
)
def dependencies_cmd(
    path: str, gitnexus_dir: str | None, output_json_flag: bool, max_targets: int
) -> None:
    """List dependency edges worth strengthening via balanced Forman curvature (#84)."""
    from topos.functors.probes.mdg.curvature import calculate_mdg_curvature
    from topos.mcp.evaluation import load_dep_graph, resolve_gitnexus_dir
    from topos.mcp.schemas import RefactorHotspot

    # An explicit --gitnexus-dir is trusted as-is (already validated by
    # click.Path(exists=True)); auto-detection (no override) falls back to
    # resolve_gitnexus_dir's "<cwd>/.gitnexus if it exists" convention.
    resolved_gitnexus_dir = (
        Path(gitnexus_dir).expanduser().resolve()
        if gitnexus_dir
        else resolve_gitnexus_dir(None, Path.cwd())
    )
    mdg = load_dep_graph(resolved_gitnexus_dir, path)
    if mdg is None:
        click.echo(
            "No .gitnexus dependency graph found. Run 'topos depgraph generate' "
            "first, or pass --gitnexus-dir.",
            err=True,
        )
        raise SystemExit(1)

    file_id = mdg.file_node_id()
    if file_id is None:
        click.echo(f"'{path}' was not found in the dependency graph.", err=True)
        raise SystemExit(1)

    curvature = calculate_mdg_curvature(mdg, file_id)
    ranked = curvature.edges[:max_targets]
    hotspots = [
        RefactorHotspot(
            kind="dependency_edge",
            label=f"{source} -> {target}",
            filepath=path,
            score=score,
            suggestion=(
                "Highly negative curvature: many otherwise-unrelated modules "
                "route coupling through this edge. Consider extracting a "
                "shared interface to strengthen it."
                if score < 0
                else "Well-supported dependency edge; no action needed."
            ),
        )
        for source, target, score in ranked
    ]
    _print_hotspots("Dependency hotspots:", hotspots, output_json_flag)


@refactor.command("process")
@click.argument("path", type=click.Path(exists=True, dir_okay=False))
@click.option(
    "--gitnexus-dir",
    type=click.Path(exists=True, file_okay=False),
    default=None,
    help="Path to a .gitnexus/ directory (auto-detected from cwd when omitted).",
)
@click.option(
    "--json", "output_json_flag", is_flag=True, help="Output results as JSON."
)
@click.option(
    "--max-targets",
    default=5,
    show_default=True,
    type=int,
    help="Maximum process hotspots to display.",
)
def process_cmd(
    path: str, gitnexus_dir: str | None, output_json_flag: bool, max_targets: int
) -> None:
    """List process-graph choke points via directed Forman-Ricci curvature (#86)."""
    from topos.functors.probes.process.curvature import calculate_process_curvature
    from topos.graphs.process.object import ProcessGraph
    from topos.mcp.evaluation import load_dep_graph, resolve_gitnexus_dir
    from topos.mcp.schemas import RefactorHotspot

    # An explicit --gitnexus-dir is trusted as-is (already validated by
    # click.Path(exists=True)); auto-detection (no override) falls back to
    # resolve_gitnexus_dir's "<cwd>/.gitnexus if it exists" convention.
    resolved_gitnexus_dir = (
        Path(gitnexus_dir).expanduser().resolve()
        if gitnexus_dir
        else resolve_gitnexus_dir(None, Path.cwd())
    )
    mdg = load_dep_graph(resolved_gitnexus_dir, path)
    if mdg is None:
        click.echo(
            "No .gitnexus dependency graph found. Run 'topos depgraph generate' "
            "first, or pass --gitnexus-dir.",
            err=True,
        )
        raise SystemExit(1)

    file_id = mdg.file_node_id()
    if file_id is None:
        click.echo(f"'{path}' was not found in the dependency graph.", err=True)
        raise SystemExit(1)

    process_graph = ProcessGraph.from_mdg(mdg, path)
    touching = process_graph.paths_touching_file(file_id)
    subgraph = ProcessGraph(target_file=path, paths=touching)
    curvature = calculate_process_curvature(subgraph)

    ranked = curvature.edges[:max_targets]
    hotspots = [
        RefactorHotspot(
            kind="process_transition",
            label=f"{source} -> {target}",
            filepath=path,
            score=score,
            suggestion=(
                "Choke point: many independent execution paths squeeze "
                "through this transition. Consider an asynchronous "
                "decoupling boundary (message queue / pub-sub)."
                if score < 0
                else "Well-distributed transition; no action needed."
            ),
        )
        for source, target, score in ranked
    ]
    _print_hotspots("Process choke points:", hotspots, output_json_flag)


def register_refactor_commands(cli_group: click.Group) -> None:
    """Attach the refactoring suite to the root CLI group."""
    cli_group.add_command(refactor)

File: crates/topos-mcp/src/tools/refactor.rs
//! Unified refactoring-guidance suite (Methods Upgrade milestone).
//!
//! One tool, three targets (ranked hotspot list + suggested action), each
//! surfacing a different structural-analysis engine:
//!
//! - `target="cycles"`: cycle-basis extraction on the CFG.
//! - `target="dependencies"`: balanced Forman curvature on the MDG.
//! - `target="process"`: directed Forman-Ricci curvature on GitNexus
//!   process graphs.
//!
//! All three are purely advisory — none of this feeds
//! SIMPLE/COMPOSABLE/SECURE scoring.

use rmcp::handler::server::wrapper::Parameters;
use rmcp::model::CallToolResult;
use rmcp::{tool, tool_router};
use topos_core::core::morphism::ProgramMorphism;
use topos_core::functors::probes::cfg::homology::calculate_cycle_basis;
use topos_core::functors::probes::mdg::curvature::calculate_mdg_curvature;
use topos_core::functors::probes::process::curvature::calculate_process_curvature;
use topos_core::graphs::process::object::ProcessGraph;

use crate::evaluation::{detect_language, load_dep_graph, resolve_gitnexus_dir};
use crate::formatting::to_tool_result;
use crate::refactor_hotspots::render_hotspots_md;
use crate::schemas::{RefactorHotspot, RefactorInput, RefactorResult, RefactorTargetKind};
use crate::security::{read_safe_utf8_file, resolve_file_root, resolve_within_root};
use crate::server::ToposServer;

fn err_result(target: RefactorTargetKind, filepath: &str, error: String) -> RefactorResult {
    RefactorResult {
        target: target.as_str().to_string(),
        filepath: filepath.to_string(),
        betti_1: None,
        gitnexus_available: None,
        hotspots: Vec::new(),
        error: Some(error),
    }
}

fn refactor_cycles(params: &RefactorInput) -> (RefactorResult, String) {
    let resolved = match resolve_within_root(&params.filepath) {
        Ok(path) => path,
        Err(err) => {
            let model = err_result(RefactorTargetKind::Cycles, &params.filepath, err);
            return (model, render_hotspots_md("Cycle hotspots", &[]));
        }
    };
    let source = match read_safe_utf8_file(&resolved.to_string_lossy()) {
        Ok(source) => source,
        Err(err) => {
            let model = err_result(RefactorTargetKind::Cycles, &params.filepath, err);
            return (model, render_hotspots_md("Cycle hotspots", &[]));
        }
    };

    let language = detect_language(&resolved);
    let mut morphism = ProgramMorphism::new(&source, language);
    let Some(cfg) = morphism.build_cfg().cloned() else {
        let model = err_result(
            RefactorTargetKind::Cycles,
            &params.filepath,
            "Could not build a control-flow graph.".to_string(),
        );
        return (model, render_hotspots_md("Cycle hotspots", &[]));
    };

    let result = calculate_cycle_basis(&cfg);
    let mut ranked = result.cycles.clone();
    ranked.sort_by_key(|b| std::cmp::Reverse(span(b)));
    ranked.truncate(params.limit);

    let hotspots: Vec<RefactorHotspot> = ranked
        .iter()
        .map(|cycle| RefactorHotspot {
            kind: "cycle".to_string(),
            label: format!("cycle over blocks {:?}", cycle.block_ids),
            filepath: params.filepath.clone(),
            line_start: cycle.start_line,
            line_end: cycle.end_line,
            score: span(cycle) as f64,
            suggestion: "Extract this loop/branch body into its own function to isolate the \
                         cycle and shrink cyclomatic complexity."
                .to_string(),
        })
        .collect();

    let title = format!("Cycle hotspots (betti_1={})", result.betti_1);
    let md = render_hotspots_md(&title, &hotspots);
    let model = RefactorResult {
        target: "cycles".to_string(),
        filepath: params.filepath.clone(),
        betti_1: Some(result.betti_1),
        gitnexus_available: None,
        hotspots,
        error: None,
    };
    (model, md)
}

fn span(cycle: &topos_core::functors::probes::cfg::homology::SourceCycle) -> usize {
    match (cycle.start_line, cycle.end_line) {
        (Some(start), Some(end)) => end.saturating_sub(start),
        _ => 0,
    }
}

fn refactor_dependencies(params: &RefactorInput) -> (RefactorResult, String) {
    let project_root = match resolve_file_root() {
        Ok(root) => root,
        Err(err) => {
            let model = err_result(RefactorTargetKind::Dependencies, &params.filepath, err);
            return (model, render_hotspots_md("Dependency hotspots", &[]));
        }
    };
    let gitnexus_dir = resolve_gitnexus_dir(params.gitnexus_dir.as_deref(), &project_root);
    let (mdg, _) = load_dep_graph(gitnexus_dir.as_deref(), &params.filepath);
    let Some(mdg) = mdg else {
        let model = RefactorResult {
            target: "dependencies".to_string(),
            filepath: params.filepath.clone(),
            betti_1: None,
            gitnexus_available: Some(false),
            hotspots: Vec::new(),
            error: None,
        };
        return (model, render_hotspots_md("Dependency hotspots", &[]));
    };

    let Some(file_id) = mdg.file_node_id().map(str::to_string) else {
        let model = RefactorResult {
            target: "dependencies".to_string(),
            filepath: params.filepath.clone(),
            betti_1: None,
            gitnexus_available: Some(true),
            hotspots: Vec::new(),
            error: None,
        };
        return (model, render_hotspots_md("Dependency hotspots", &[]));
    };

    let curvature = calculate_mdg_curvature(&mdg, &file_id, "IMPORTS");
    let hotspots: Vec<RefactorHotspot> = curvature
        .edges
        .iter()
        .take(params.limit)
        .map(|(src, dst, score)| RefactorHotspot {
            kind: "dependency_edge".to_string(),
            label: format!("{src} -> {dst}"),
            filepath: params.filepath.clone(),
            line_start: None,
            line_end: None,
            score: *score,
            suggestion: if *score < 0.0 {
                "Highly negative curvature: many otherwise-unrelated modules route coupling \
                 through this edge. Consider extracting a shared interface or reducing the \
                 shared surface to strengthen it."
                    .to_string()
            } else {
                "Well-supported dependency edge; no action needed.".to_string()
            },
        })
        .collect();

    let md = render_hotspots_md("Dependency hotspots", &hotspots);
    let model = RefactorResult {
        target: "dependencies".to_string(),
        filepath: params.filepath.clone(),
        betti_1: None,
        gitnexus_available: Some(true),
        hotspots,
        error: None,
    };
    (model, md)
}

fn refactor_process(params: &RefactorInput) -> (RefactorResult, String) {
    let project_root = match resolve_file_root() {
        Ok(root) => root,
        Err(err) => {
            let model = err_result(RefactorTargetKind::Process, &params.filepath, err);
            return (model, render_hotspots_md("Process choke points", &[]));
        }
    };
    let gitnexus_dir = resolve_gitnexus_dir(params.gitnexus_dir.as_deref(), &project_root);
    let (mdg, _) = load_dep_graph(gitnexus_dir.as_deref(), &params.filepath);
    let Some(mdg) = mdg else {
        let model = RefactorResult {
            target: "process".to_string(),
            filepath: params.filepath.clone(),
            betti_1: None,
            gitnexus_available: Some(false),
            hotspots: Vec::new(),
            error: None,
        };
        return (model, render_hotspots_md("Process choke points", &[]));
    };

    let Some(file_id) = mdg.file_node_id().map(str::to_string) else {
        let model = RefactorResult {
            target: "process".to_string(),
            filepath: params.filepath.clone(),
            betti_1: None,
            gitnexus_available: Some(true),
            hotspots: Vec::new(),
            error: None,
        };
        return (model, render_hotspots_md("Process choke points", &[]));
    };

    let process_graph = ProcessGraph::from_mdg(&mdg, params.filepath.clone());
    let touching = process_graph.paths_touching_file(&file_id);
    let subgraph = ProcessGraph::from_paths(params.filepath.clone(), touching);
    let curvature = calculate_process_curvature(&subgraph, None);

    let hotspots: Vec<RefactorHotspot> = curvature
        .edges
        .iter()
        .take(params.limit)
        .map(|(src, dst, score)| RefactorHotspot {
            kind: "process_transition".to_string(),
            label: format!("{src} -> {dst}"),
            filepath: params.filepath.clone(),
            line_start: None,
            line_end: None,
            score: *score,
            suggestion: if *score < 0.0 {
                "Choke point: many independent execution paths squeeze through this \
                 transition. Consider an asynchronous decoupling boundary (message queue / \
                 pub-sub), or acknowledge the simplicity trade-off of keeping it."
                    .to_string()
            } else {
                "Well-distributed transition; no action needed.".to_string()
            },
        })
        .collect();

    let md = render_hotspots_md("Process choke points", &hotspots);
    let model = RefactorResult {
        target: "process".to_string(),
        filepath: params.filepath.clone(),
        betti_1: None,
        gitnexus_available: Some(true),
        hotspots,
        error: None,
    };
    (model, md)
}

#[tool_router(router = refactor_router, vis = "pub(crate)")]
impl ToposServer {
    /// Refactor hotspots (read-only). Three targets: `cycles` (CFG cycle
    /// basis pointing at loop bodies), `dependencies` (MDG Forman curvature
    /// naming load-bearing import edges), `process` (process-graph choke
    /// points). All purely advisory — none feeds the lattice score.
    #[tool(
        name = "topos_refactor",
        annotations(
            title = "Topos Refactor Suggestions",
            read_only_hint = true,
            destructive_hint = false,
            idempotent_hint = true,
            open_world_hint = false
        )
    )]
    pub fn topos_refactor(&self, Parameters(params): Parameters<RefactorInput>) -> CallToolResult {
        let (model, md) = match params.target {
            RefactorTargetKind::Cycles => refactor_cycles(&params),
            RefactorTargetKind::Dependencies => refactor_dependencies(&params),
            RefactorTargetKind::Process => refactor_process(&params),
        };
        to_tool_result(&model, md)
    }
}
